src/toxcore/dht/dht_friend.rs

```diff
@@ -0,0 +1,381 @@
+/*
+    Copyright (C) 2013 Tox project All Rights Reserved.
+    Copyright © 2018 Namsoo CHO <nscho66@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*!
+Module for friend.
+*/
+
+use std::time::{Duration, Instant};
+use std::io::{Error, ErrorKind};
+use std::mem;
+
+use futures::{future, Future, stream, Stream};
+
+use toxcore::dht::packed_node::*;
+use toxcore::dht::kbucket::*;
+use toxcore::crypto_core::*;
+use toxcore::dht::server::*;
+use toxcore::dht::server::client::*;
+use toxcore::io_tokio::IoFuture;
+
+/// Hold friend related info.
+pub struct DhtFriend {
+    // Public key of friend
+    pk: PublicKey,
+    // close nodes of friend
+    close_nodes: Bucket,
+    // Last time of NodesRequest packet sent
+    last_nodes_req_time: Instant,
+    // Counter for bootstappings.
+    bootstrap_times: u32,
+    // Nodes to bootstrap.
+    bootstrap_nodes: Bucket,
+}
+
+impl DhtFriend {
+    /// Create new DhtFriend object
+    /// Maximum bootstrap_times is 5, if you want to bootstrap 2 times, set bootstrap_times to 3.
+    pub fn new(pk: PublicKey, bootstrap_times: u32) -> Self {
+        DhtFriend {
+            pk,
+            close_nodes: Bucket::new(None),
+            last_nodes_req_time: Instant::now(),
+            bootstrap_times,
+            bootstrap_nodes: Bucket::new(None),
+        }
+    }
+
+    /// send NodesRequest packet to bootstap_nodes, close list
+    pub fn send_nodes_req_packets(&mut self, server: &Server,
+                                  ping_interval: u64, nodes_req_interval: u64, bad_node_timeout: u64) -> IoFuture<()> {
+        let ping_bootstrap_nodes = self.ping_bootstrap_nodes(server);
+        let ping_and_get_close_nodes = self.ping_and_get_close_nodes(server, ping_interval);
+        let send_nodes_req_random = self.send_nodes_req_random(server, bad_node_timeout, nodes_req_interval);
+
+        let res = ping_bootstrap_nodes.join3(
+            ping_and_get_close_nodes, send_nodes_req_random
+            ).map(|((),(),())| () );
+
+        Box::new(res)
+    }
+
+    fn ping_bootstrap_nodes(&mut self, server: &Server) -> IoFuture<()> {
+        let mut bootstrap_nodes = Bucket::new(None);
+        mem::swap(&mut bootstrap_nodes, &mut self.bootstrap_nodes);
```

Comment:
Better use Option<Bucket> for DhtFriend::bootstrap_nodes

Reply:
You sure that option is better? In this case we will have to check if it's Some or if it's None when we add new node to the list. If Some - add new node, if None - make it Some and again add new node. Seems to me like one additional if for no reason - empty bucket just means that there is no nodes.
In fact this `mem::swap` was added to mitigate borrow chacker and not to obtain lock twice (in the same code in the Server struct).

Reply:
OK